### DISCUSSION TITLE
Isolate .cargo directory inside .pixi directory

### DISCUSSION THREAD

**Discussion by jonmmease** [prefix-dev/pixi#275] [Ideas]
Hi, I'm working on moving my [VegaFusion](https://github.com/hex-inc/vegafusion/) project to pixi. This is a project with a Rust core and wrappers in Python, wasm / JavaScript, and Java. The current development setup is a bit of a hassle involving the system installation of rust/cargo, node, protobuf, and conda/conda.lock. I'd like to see if I can isolate all of this with pixi, as conda-forge already provides rust/cargo, node, and protobuf.

This seems to be working out pretty well so far (it's still a WIP), but I had one idea that I think would improve the situation a bit.  I noticed that the rust installation from conda-forge still creates a `~/.cargo` directory for the cargo cache and it places any globally installed executables in `~/.cargo/bin`. It would be great to have a clean way to put this `.cargo` directory inside the `.pixi` folder, and to add the `.cargo/bin` directory to the path for use in tasks and in `pixi shell`. This would make it possible to build a rust project without a system installation of rust, and without leaving any traces of Rust configuration outside of the project's pixi directory.

As a side note, the reason I need `.cargo/bin` on the path is because I'm using `cargo` to install `wasm-pack`, which doesn't seem to be available on conda-forge at this point.

Here's one approach that comes to mind. It's possible to move the `.cargo` directory by setting the `CARGO_HOME` environment variable, so if there was a way to set environment variables in `pixi.toml` that would be used across tasks and in `pixi shell`, that would make it possible to move the `.cargo` directory to `.pixi/.cargo`.  And if it were possible to add additional directories inside `.pixi` to the PATH, then it would be possible to add `.pixi/.cargo/bin` to the path for use in tasks and `pixi shell`.

**Comment by wolfv** [comment]
Hey @jonmmease, I think we have two options:

- we could add a `CARGO_HOME` variable to the activation script of the Rust feedstock (that would change this behavior for everyone though)
- we are working on adding environment variable support to pixi run, so that you could set / override whatever env vars you'd like

I am not sure if everyone that uses the rust feedstock would prefer to _not_ share their cargo cache ... so I am a bit unsure. We'll for sure work on adding environment variable support in pixi though.

**Reply by jonmmease** [reply]
Thanks for the response. I think I'd prefer the environment variable approach (and glad to hear you're working on it!), as there's probably an argument for keeping`.cargo` in the default location in order to share the cache across conda environments in certain situations.

**Reply by twrightsman** [reply]
I think setting `CARGO_HOME` in the feedstock is actually a good idea as well, because then Conda environments are more isolated from the system and each other, as they should be?
